### PR TITLE
Improve emoji UI and ability to react by clicking on a reaction

### PR DESCRIPTION
**Pull request by Radu-Raicea**
1. Reactions are now under the chat bubble.
2. You can react by clicking on an existing reaction. If you have already reacted with the emoji you clicked on, it will show a toast telling you that you can't react again.
3. Duplicate reactions are displayed only once with their count next to them.
4. Long clicking a reaction shows a dialog with the number of people that reacted with that emoji, each sender's name and the date.

### CODE REVIEW

File: src/org/thoughtcrime/securesms/ConversationItem.java [hunk 1]
```diff
@@ -533,20 +537,57 @@ protected Void doInBackground(Void... params) {
...
+            return;
+          }
+        }
+
+        try {
+          ((ConversationActivity) context).handleNewReaction(messageRecord, new ReactionsHandler(getContext()), reaction.getReaction());
+        } catch (InvalidMessageException e) {
+          e.printStackTrace();
+        }
+
```
Comment (veken1199) [inline src/org/thoughtcrime/securesms/ConversationItem.java]: extra line

File: src/org/thoughtcrime/securesms/database/MessageReactionDatabase.java [hunk 2]
```diff
@@ -67,23 +67,20 @@ public void reactToMessage(MessageRecord messageRecord, String reaction, Long re
...
+        String address;
+
+        try {
+            address = DatabaseFactory.getIdentityDatabase(context).getMyIdentity().getAddress().serialize();
+        } catch (Exception e) {
+            e.printStackTrace();
+            Parcel p = Parcel.obtain();
+            p.writeString("Me");
+            p.setDataPosition(0);
+            address = new Address(p).serialize();
```
Comment (veken1199) [inline src/org/thoughtcrime/securesms/database/MessageReactionDatabase.java:79]: +1 for adding serialize

**Review by veken1199 — DISMISSED** [review]
Looks good

**Comment by erinb90** [comment]
@Radu-Raicea could you post a screenshot of the UI changes?
Edit: I posted some.

File: res/layout/conversation_item_sent.xml [hunk 3]
```diff
@@ -197,6 +188,17 @@
...
+            android:id="@+id/reactions_list"
+            android:layout_width="200sp"
+            android:layout_height="wrap_content"
+            android:layout_alignRight="@+id/body_bubble"
+            android:layout_below="@+id/body_bubble"
+            android:gravity="right"
+            android:orientation="horizontal"
+            app:flexWrap="wrap" />
+
+
```
Comment (erinb90) [inline res/layout/conversation_item_sent.xml]: extra line

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 4]
```diff
@@ -821,6 +798,44 @@ public void onEmojiSelected(String emoji) throws InvalidMessageException {
...
+    inputPanel.setEmojiDrawer(emojiDrawerStub.get());
+    emojiDrawerStub.get().setEmojiEventListener(inputPanel);
+    inputPanel.setVisibility(View.VISIBLE);
+    linMessage.setVisibility(View.VISIBLE);
+  }
+
+  public void handleNewReaction(MessageRecord messageRecord, ReactionsHandler handler, String emoji) throws InvalidMessageException {
+    Long time = currentTimeMillis();
+
+    //insert new reaction into db
```
Comment (erinb90) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: I think this comment should be removed, the method name is self explanatory

File: src/org/thoughtcrime/securesms/ConversationActivity.java [hunk 5]
```diff
@@ -821,6 +798,44 @@ public void onEmojiSelected(String emoji) throws InvalidMessageException {
...
+    Map<String, String> map = new HashMap<>();
+    map.put("type", "reaction");
+    map.put("hash", messageRecord.getHash());
+    map.put("emoji", emoji);
+    map.put("time", time.toString());
+    map.put("address", myAddress);
+
+    String body = new JSONObject(map).toString();
+
+    if(recipient.getAddress().isGroup()) {
```
Comment (erinb90) [inline src/org/thoughtcrime/securesms/ConversationActivity.java]: missing space after if

File: src/org/thoughtcrime/securesms/ConversationItem.java [hunk 6]
```diff
@@ -533,20 +537,57 @@ protected Void doInBackground(Void... params) {
   }
 
   private void setReactions(final MessageRecord messageRecord) {
-    LinearLayout reactionsList = (LinearLayout) findViewById(R.id.reactions_list);
+    FlexboxLayout reactionsList = (FlexboxLayout) findViewById(R.id.reactions_list);
 
-    if(((LinearLayout) reactionsList).getChildCount() > 0) {
-      ((LinearLayout) reactionsList).removeAllViews();
+    if(((FlexboxLayout) reactionsList).getChildCount() > 0) {
```
Comment (erinb90) [inline src/org/thoughtcrime/securesms/ConversationItem.java]: missing space after if

File: src/org/thoughtcrime/securesms/ConversationItem.java [hunk 7]
```diff
@@ -533,20 +537,57 @@ protected Void doInBackground(Void... params) {
...
+
+        String address = "";
+
+        try {
+          address = DatabaseFactory.getIdentityDatabase(context).getMyIdentity().getAddress().toString();
+        } catch (Exception e) {
+          e.printStackTrace();
+        }
+
+        for(ReactionsHandler.Reaction reaction_check : reactions) {
```
Comment (erinb90) [inline src/org/thoughtcrime/securesms/ConversationItem.java]: missing space after for

**Review by shelbaz — DISMISSED** [review]
Tested - Looks good!

**Comment by erinb90** [comment]
Screenshot of new UI: https://user-images.githubusercontent.com/14361988/38518314-8ec57ec0-3c0a-11e8-99be-bc34afd3f27f.png

**Comment by Radu-Raicea** [comment]
@erinb90 @veken1199 
I've made your code review changes.

**Comment by erinb90** [comment]
Screenshot of long clicking reaction: https://user-images.githubusercontent.com/14361988/38524260-03378a1c-3c1c-11e8-9526-8a441e6ca987.png

**Review by erinb90 — APPROVED** [review]
Looks good

**Review by shelbaz — APPROVED** [review]
Looks good !

**Review by veken1199 — APPROVED** [review]

**Pull request merged**